include/swift/SIL/OwnershipUtils.h

```diff
@@ -769,13 +770,97 @@ inline AddressUseKind meet(AddressUseKind lhs, AddressUseKind rhs) {
   return (lhs > rhs) ? lhs : rhs;
 }
 
+/// A state structure for findTransitiveUsesForAddress. Intended to be only used
+/// a single time. Please always use a new one for every call to
+/// findTransitiveUsesForAddress.
```

Comment:
I don't want to introduce a unique_ptr or a moveable rvalue method (mainly b/c even though we use that in parts of the code base, it is a foreign feature to many users).

I think a better pattern would be to make the result Optional and ban the copy constructor. Then when findTransitiveUsesForAddress completes, we invalidate the result. Then if someone tries to call it again, they get a run time crash.

Reply:
A runtime crash is not better than a static check by the compiler that the API is used correctly. The only users of this API are Swift compiler developers, who have to understand movable rvalue methods as we use it elsewhere.

Reply:
@kavon I disagree. If someone misuses this API and get a runtime crash, it is going to happen immediately when they use the API... that suggests either way someone will get a quick slap on the wrist showing them what the issue is. I think this is a matter of personal style and taste.

Reply:
> If someone misuses this API and get a runtime crash, it is going to happen immediately when they use the API

By the nature of runtime tests, this is not guaranteed. The following gets statically rejected by an rvalue-method version, but may not be caught by testing and would be a subtle bug:

```
MyVisitor visitor;
findTransitiveUsesForAddress(address, visitor);
// ...
if (uncommonScenario)
  findTransitiveUsesForAddress(address, visitor);
```

This is not a matter of taste. Relying on a runtime test to enforce lifetime rules is strictly less robust and mistake-prone. Please make use of C++'s lifetime features as we do elsewhere in the compiler to prevent mistakes.